### PR TITLE
[REST] URL Encoder and Routes

### PR DESCRIPTION
**Pull request by abuneri**
This is adding a `util::url_encode` method for all reaction-related endpoints. 
Now users can just use codepoints, "U\XXXXXXXX",  or the unicode emoji itself to use these endpoints.
Also, a few routes needed a touch up.

**NOTE:** this has been tested to work with the following cases:
1. "U\XXXXXXXX" - **codepoints**
2. "✅" - **unicode emoji itself**
3. "disccord_memes:1234567891012313" - **custom emojis**

Only does not work when user sends in the percent encoding themselves, (e.g. "%EA%B4%FF"). But this is a desired abstraction.

### Example Usage in the Library (Encoding the URL)
```cpp
// these methods are used when generating the reaction-related routes.
util::url_encode("\U00002705"); //output: %E2%9C%85 (White Heavy Check Mark)
util::url_encode("✅");         //output: %E2%9C%85 (White Heavy Check Mark)
util::url_encode("\U0001F914"); //output: %F0%9F%A4%94 (Thinking Face)
util::url_encode("\U00002B50"); //output: %E2%AD%90 (White Medium Star)
```

### Example Usage for a User (Message Reacting)
```cpp
// sending a message then auto reacting w/ heavy check mark

uint64_t c_id = 1234567891011;
create_message_args cm_args("wow disccord is such an amazing lib!");
auto msg = client.create_message(c_id, cm_args).get();

// using codepoints
client.create_reaction(msg.get_channel_id(), msg.get_id(), "\U00002705").wait();

// or using actual emoji
client.create_reaction(msg.get_channel_id(), msg.get_id(), "✅").wait();
```

### CODE REVIEW

**Comment by abuneri** [comment]
@FiniteReality good2go

File: lib/rest/api_client.cpp [hunk 1]
```diff
@@ -255,25 +257,25 @@ namespace disccord
 
             pplx::task<void> rest_api_client::create_reaction(uint64_t channel_id, uint64_t message_id, std::string emoji, const pplx::cancellation_token& token)
             {
-                auto route = get_route("PUT", "/channels/{channel.id}/messages/{message.id}/reactions/{emoji}/@me", std::to_string(channel_id), std::to_string(message_id), emoji);
+                auto route = get_route("PUT", "/channels/{channel.id}/messages/{message.id}/reactions/"+util::url_encode(emoji)+"/@me", std::to_string(channel_id), std::to_string(message_id));
```
Comment (FiniteReality) [inline lib/rest/api_client.cpp:260]: This didn't really need to change much, did it? Just wrap the original emoji parameter in `util::url_encode`?

File: lib/rest/api_client.cpp [hunk 2]
```diff
@@ -255,25 +257,25 @@ namespace disccord
...
                 return request(route, token);
             }
 
             pplx::task<void> rest_api_client::delete_own_reaction(uint64_t channel_id, uint64_t message_id, std::string emoji, const pplx::cancellation_token& token)
             {
-                auto route = get_route("DELETE", "/channels/{channel.id}/messages/{message.id}/reactions/{emoji}/@me", std::to_string(channel_id), emoji, std::to_string(message_id));
+                auto route = get_route("DELETE", "/channels/{channel.id}/messages/{message.id}/reactions/"+util::url_encode(emoji)+"/@me", std::to_string(channel_id), std::to_string(message_id));
```
Comment (FiniteReality) [inline lib/rest/api_client.cpp:266]: Likewise, just wrap the original parameter

File: lib/rest/api_client.cpp [hunk 3]
```diff
@@ -255,25 +257,25 @@ namespace disccord
...
                 return request(route, token);
             }
 
             pplx::task<void> rest_api_client::delete_user_reaction(uint64_t channel_id, uint64_t message_id, uint64_t user_id, std::string emoji, const pplx::cancellation_token& token)
             {
-                auto route = get_route("DELETE", "/channels/{channel.id}/messages/{message.id}/reactions/{emoji}/{user_id}", std::to_string(channel_id), std::to_string(message_id), emoji, std::to_string(user_id));
+                auto route = get_route("DELETE", "/channels/{channel.id}/messages/{message.id}/reactions/"+util::url_encode(emoji)+"/{user.id}", std::to_string(channel_id), std::to_string(message_id), std::to_string(user_id));
```
Comment (FiniteReality) [inline lib/rest/api_client.cpp:272]: Here too, just wrap the emoji parameter.

**Review by FiniteReality — CHANGES_REQUESTED** [review]

File: lib/rest/api_client.cpp (hunk 1 above)
Comment (abuneri) [reply lib/rest/api_client.cpp:260]: see #30 for why i needed to do it like this for now.

File: lib/rest/api_client.cpp (hunk 2 above)
Comment (abuneri) [reply lib/rest/api_client.cpp:266]: #30

File: lib/rest/api_client.cpp (hunk 3 above)
Comment (abuneri) [reply lib/rest/api_client.cpp:272]: #30

**Comment by abuneri** [comment]
since routing and stuff changed in feature/websocket, i'm making a new PR for `url_encode` for that branch instead, closing this now.

**Pull request closed without merging**